### PR TITLE
Support json deserialize of 'null' and 'bool' when declared type is typeof(object)

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/corefx/issues/36611 which previously threw a NullReferenceException instead of returning null when deserializing a null root object that has typeof(object):
```object obj = JsonSerializer.Parse<object>("null");```

Also includes refactoring around polymorphic support and enables someone to write an object-based value converter in the future.

### CODE REVIEW

File: src/System.Text.Json/tests/Serialization/String.ReadTests.cs [hunk 1]
```diff
@@ -1,53 +0,0 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
-
-using Xunit;
-
-namespace System.Text.Json.Serialization.Tests
-{
-    public static partial class StringTests
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/String.ReadTests.cs]: Are we removing any test coverage here or did all of them moved to `ObjectTests`?

**Comment by ahsonkhan** [comment]
> Also adds support for polymorphic deserialization of a bool when the declared type is typeof(object). Bools are the only type that can be deserialized (besides null) like this since the token type is not ambiguous (unlike strings, numbers and arrays which can map to many types). For other types we plan on deserializing to JsonElement and arrays in the future.

Should we be singling out bools like this? Why not have them return JsonElement too?

**Comment by ahsonkhan** [comment]
https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F36637~2Fmerge/test~2Ffunctional~2Fcli~2F/20190405.25/workItem/System.Text.Json.Tests/analysis/xunit/System.Text.Json.Serialization.Tests.NullTests~2FNullReferences
 System.Text.Json.Serialization.Tests.NullTests/NullReferences
```text
Unhandled Exception of Type Xunit.Sdk.EqualException
Message :
Assert.Equal() Failure
                                 ↓ (pos 61)
Expected: ···\"NullableInt\":null,\"Bool\":true,\"Object\":null,\"NullableIntArra···
Actual:   ···\"NullableInt\":null,\"NullableIntArray\":null,\"Bool\":true,\"Objec···
                                 ↑ (pos 61)
Stack Trace :
   at System.Text.Json.Serialization.Tests.NullTests.NullReferences() in D:\a\1\s\src\System.Text.Json\tests\Serialization\Null.WriteTests.cs:line 42
```

File: src/System.Text.Json/tests/Serialization/String.ReadTests.cs (hunk 1 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/String.ReadTests.cs]: Just refactoring. All tests still there.

**Comment by steveharter** [comment]
> Should we be singling out bools like this? Why not have them return JsonElement too?

I mentioned in the intro section that bools are special because they are unambiguous (token type True and False). However, if we want consistency across `JsonElement` then yes perhaps I should throw NotSupportedException here as well instead (and leave that decision until later).

File: src/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs [hunk 2]
```diff
@@ -21,6 +21,12 @@ public static void ThrowJsonReaderException_DeserializeUnableToConvertValue(Type
             throw new JsonReaderException(SR.Format(SR.DeserializeUnableToConvertValue, state.PropertyPath, propertyType), reader.CurrentState);
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public static void ThrowJsonReaderException_DeserializeUnableToConvertValue(Type propertyType, in ReadStack state)
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs]: Where is this called from?

**Comment by ahsonkhan** [comment]
Merging to unblock upstack consumption. cc @BrennanConroy

**Pull request merged**